File: services/frontend/lib/ai.ts
import { fetchApi } from './api';

export interface AIProviderSummary {
  key: string;
  type: string;
  label: string;
  chatModel: string;
  baseUrl?: string;
  default: boolean;
  tokenConfigured: boolean;
}

export interface AIProviderAdminSettings {
  providerKey: string;
  providerType: string;
  label: string;
  baseUrl: string;
  apiPath: string;
  apiVersion: string;
  region: string;
  organization: string;
  chatModel: string;
  embeddingModel: string;
  enabled: boolean;
  isDefault: boolean;
  configured: boolean;
  tokenConfigured: boolean;
  requiresToken: boolean;
  timeoutSeconds: number;
  maxContextTokens: number;
  maxOutputTokens: number;
  temperature: number;
}

export interface AIMessageSource {
  appId: string;
  appName: string;
  chunkId: string;
  sourceType: string;
  sourceId: string;
  title: string;
  snippet: string;
  score?: number;
}

export interface AIMessage {
  id: string;
  conversationId: string;
  role: 'user' | 'assistant' | 'system' | 'tool';
  content: string;
  providerKey?: string;
  providerType?: string;
  model?: string;
  promptTokens?: number;
  responseTokens?: number;
  sources?: AIMessageSource[];
  error?: string;
  createdAt: string;
}

export interface AIConversation {
  id: string;
  userId: string;
  title: string;
  scopeType: string;
  appId?: string;
  createdAt: string;
  updatedAt: string;
  messages?: AIMessage[];
}

export interface AISendMessageResponse {
  conversation: AIConversation;
  userMessage: AIMessage;
  assistantMessage: AIMessage;
  sources: AIMessageSource[];
}

export interface AISendMessagePayload {
  conversationId?: string;
  message: string;
  appId?: string;
  providerKey?: string;
}

async function parseError(response: Response, fallback: string): Promise<Error> {
  const data = await response.json().catch(() => ({})) as { message?: string; error?: string };
  const message = data.message?.trim();
  const detail = data.error?.trim();
  if (message && detail && detail !== message) return new Error(`${message}: ${detail}`);
  return new Error(message || detail || fallback);
}

export async function listAIProviders(): Promise<AIProviderSummary[]> {
  const response = await fetchApi('/ai/providers', { cache: 'no-store' });
  if (!response.ok) throw await parseError(response, 'AI-Provider konnten nicht geladen werden.');
  const data = await response.json();
  return Array.isArray(data) ? data : [];
}

export async function listAIConversations(): Promise<AIConversation[]> {
  const response = await fetchApi('/ai/conversations', { cache: 'no-store' });
  if (!response.ok) throw await parseError(response, 'Konversationen konnten nicht geladen werden.');
  const data = await response.json();
  return Array.isArray(data) ? data : [];
}

export async function getAIConversation(id: string): Promise<AIConversation> {
  const response = await fetchApi(`/ai/conversations/${id}`, { cache: 'no-store' });
  if (!response.ok) throw await parseError(response, 'Konversation konnte nicht geladen werden.');
  return response.json();
}

export async function deleteAIConversation(id: string): Promise<void> {
  const response = await fetchApi(`/ai/conversations/${id}`, { method: 'DELETE' });
  if (!response.ok) throw await parseError(response, 'Konversation konnte nicht gelöscht werden.');
}

export async function sendAIMessage(payload: AISendMessagePayload): Promise<AISendMessageResponse> {
  const endpoint = payload.conversationId
    ? `/ai/conversations/${payload.conversationId}/messages`
    : '/ai/chat';
  const response = await fetchApi(endpoint, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw await parseError(response, 'AI-Antwort konnte nicht erzeugt werden.');
  return response.json();
}

export async function listAIProviderSettings(): Promise<AIProviderAdminSettings[]> {
  const response = await fetchApi('/settings/ai-providers', { cache: 'no-store' });
  if (!response.ok) throw await parseError(response, 'AI-Provider konnten nicht geladen werden.');
  const data = await response.json();
  return Array.isArray(data) ? data : [];
}

export async function reindexAIKnowledge(): Promise<{ indexedApps: number }> {
  const response = await fetchApi('/ai/knowledge/reindex', { method: 'POST' });
  if (!response.ok) throw await parseError(response, 'AI-Wissensindex konnte nicht aufgebaut werden.');
  return response.json();
}

export const AI_PROVIDER_TYPES = [
  'openai',
  'azure-openai',
  'anthropic',
  'gemini',
  'mistral',
  'cohere',
  'ollama',
  'vllm',
  'openai-compatible',
  'openrouter',
  'lmstudio',
  'together',
] as const;

export type AIProviderType = typeof AI_PROVIDER_TYPES[number];

export function aiProviderTypeLabel(providerType: string): string {
  switch (providerType) {
    case 'openai':
      return 'OpenAI';
    case 'azure-openai':
      return 'Azure OpenAI';
    case 'anthropic':
      return 'Anthropic';
    case 'gemini':
      return 'Google Gemini';
    case 'mistral':
      return 'Mistral';
    case 'cohere':
      return 'Cohere';
    case 'ollama':
      return 'Ollama';
    case 'vllm':
      return 'vLLM';
    case 'openrouter':
      return 'OpenRouter';
    case 'lmstudio':
      return 'LM Studio';
    case 'together':
      return 'Together AI';
    default:
      return 'OpenAI-kompatibel';
  }
}


export function defaultAIBaseUrl(providerType: string): string {
  switch (providerType) {
    case 'openai':
      return 'https://api.openai.com/v1';
    case 'anthropic':
      return 'https://api.anthropic.com';
    case 'gemini':
      return 'https://generativelanguage.googleapis.com';
    case 'mistral':
      return 'https://api.mistral.ai/v1';
    case 'cohere':
      return 'https://api.cohere.com';
    case 'ollama':
      return 'http://localhost:11434';
    case 'vllm':
      return 'http://localhost:8000/v1';
    case 'openrouter':
      return 'https://openrouter.ai/api/v1';
    case 'lmstudio':
      return 'http://localhost:1234/v1';
    case 'together':
      return 'https://api.together.xyz/v1';
    default:
      return '';
  }
}

export function defaultAIModel(providerType: string): string {
  switch (providerType) {
    case 'openai':
    case 'openai-compatible':
    case 'openrouter':
    case 'together':
      return 'gpt-4o-mini';
    case 'anthropic':
      return 'claude-3-5-haiku-latest';
    case 'gemini':
      return 'gemini-1.5-flash';
    case 'mistral':
      return 'mistral-small-latest';
    case 'cohere':
      return 'command-r';
    case 'ollama':
      return 'llama3.1';
    default:
      return 'local-model';
  }
}

File: services/backend/pkg/models/ai.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/uptrace/bun"
)

type AIProviderSettings struct {
	bun.BaseModel `bun:"table:ai_provider_settings,alias:aip"`

	ProviderKey      string    `bun:"provider_key,pk" json:"providerKey"`
	ProviderType     string    `bun:"provider_type,notnull,default:'openai-compatible'" json:"providerType"`
	Label            string    `bun:"label" json:"label"`
	BaseURL          string    `bun:"base_url" json:"baseUrl"`
	APIPath          string    `bun:"api_path" json:"apiPath"`
	APIVersion       string    `bun:"api_version" json:"apiVersion"`
	Region           string    `bun:"region" json:"region"`
	Organization     string    `bun:"organization" json:"organization"`
	ChatModel        string    `bun:"chat_model" json:"chatModel"`
	EmbeddingModel   string    `bun:"embedding_model" json:"embeddingModel"`
	EncryptedToken   string    `bun:"encrypted_token,notnull,default:''" json:"encryptedToken,omitempty"`
	TokenNonce       string    `bun:"token_nonce,notnull,default:''" json:"tokenNonce,omitempty"`
	TokenKeyVersion  string    `bun:"token_key_version,notnull,default:'v1'" json:"tokenKeyVersion,omitempty"`
	TokenConfigured  bool      `bun:"token_configured,notnull,default:false" json:"tokenConfigured"`
	Enabled          bool      `bun:"enabled,notnull,default:true" json:"enabled"`
	IsDefault        bool      `bun:"is_default,notnull,default:false" json:"isDefault"`
	TimeoutSeconds   int       `bun:"timeout_seconds,notnull,default:30" json:"timeoutSeconds"`
	MaxContextTokens int       `bun:"max_context_tokens,notnull,default:6000" json:"maxContextTokens"`
	MaxOutputTokens  int       `bun:"max_output_tokens,notnull,default:1200" json:"maxOutputTokens"`
	Temperature      float64   `bun:"temperature,notnull,default:0.2" json:"temperature"`
	CreatedAt        time.Time `bun:"created_at,notnull,default:current_timestamp" json:"createdAt"`
	UpdatedAt        time.Time `bun:"updated_at,notnull,default:current_timestamp" json:"updatedAt"`
}

type AIProviderAdminResponse struct {
	ProviderKey      string  `json:"providerKey"`
	ProviderType     string  `json:"providerType"`
	Label            string  `json:"label"`
	BaseURL          string  `json:"baseUrl"`
	APIPath          string  `json:"apiPath"`
	APIVersion       string  `json:"apiVersion"`
	Region           string  `json:"region"`
	Organization     string  `json:"organization"`
	ChatModel        string  `json:"chatModel"`
	EmbeddingModel   string  `json:"embeddingModel"`
	Enabled          bool    `json:"enabled"`
	IsDefault        bool    `json:"isDefault"`
	Configured       bool    `json:"configured"`
	TokenConfigured  bool    `json:"tokenConfigured"`
	RequiresToken    bool    `json:"requiresToken"`
	TimeoutSeconds   int     `json:"timeoutSeconds"`
	MaxContextTokens int     `json:"maxContextTokens"`
	MaxOutputTokens  int     `json:"maxOutputTokens"`
	Temperature      float64 `json:"temperature"`
}

type AIProviderSummary struct {
	Key             string `json:"key"`
	Type            string `json:"type"`
	Label           string `json:"label"`
	ChatModel       string `json:"chatModel"`
	BaseURL         string `json:"baseUrl,omitempty"`
	Default         bool   `json:"default"`
	TokenConfigured bool   `json:"tokenConfigured"`
}

type AIConversation struct {
	bun.BaseModel `bun:"table:ai_conversations,alias:aic"`

	ID        uuid.UUID   `bun:"id,pk,type:uuid,default:gen_random_uuid()" json:"id"`
	UserID    uuid.UUID   `bun:"user_id,type:uuid,notnull" json:"userId"`
	Title     string      `bun:"title,notnull,default:''" json:"title"`
	ScopeType string      `bun:"scope_type,notnull,default:''" json:"scopeType"`
	AppID     string      `bun:"app_id,nullzero" json:"appId"`
	CreatedAt time.Time   `bun:"created_at,notnull,default:current_timestamp" json:"createdAt"`
	UpdatedAt time.Time   `bun:"updated_at,notnull,default:current_timestamp" json:"updatedAt"`
	Messages  []AIMessage `bun:"rel:has-many,join:id=conversation_id" json:"messages,omitempty"`
}

type AIMessageSource struct {
	AppID      string  `json:"appId"`
	AppName    string  `json:"appName"`
	ChunkID    string  `json:"chunkId"`
	SourceType string  `json:"sourceType"`
	SourceID   string  `json:"sourceId"`
	Title      string  `json:"title"`
	Snippet    string  `json:"snippet"`
	Score      float64 `json:"score,omitempty"`
}

type AIMessage struct {
	bun.BaseModel `bun:"table:ai_messages,alias:aim"`

	ID             uuid.UUID         `bun:"id,pk,type:uuid,default:gen_random_uuid()" json:"id"`
	ConversationID uuid.UUID         `bun:"conversation_id,type:uuid,notnull" json:"conversationId"`
	Role           string            `bun:"role,notnull" json:"role"`
	Content        string            `bun:"content,notnull" json:"content"`
	ProviderKey    string            `bun:"provider_key" json:"providerKey"`
	ProviderType   string            `bun:"provider_type" json:"providerType"`
	Model          string            `bun:"model" json:"model"`
	PromptTokens   int               `bun:"prompt_tokens,notnull,default:0" json:"promptTokens"`
	ResponseTokens int               `bun:"response_tokens,notnull,default:0" json:"responseTokens"`
	Sources        []AIMessageSource `bun:"sources,type:jsonb,notnull,default:'[]'" json:"sources"`
	Error          string            `bun:"error,notnull,default:''" json:"error"`
	CreatedAt      time.Time         `bun:"created_at,notnull,default:current_timestamp" json:"createdAt"`
}

type AIKnowledgeChunk struct {
	bun.BaseModel `bun:"table:ai_knowledge_chunks,alias:aikc"`

	ID              uuid.UUID         `bun:"id,pk,type:uuid,default:gen_random_uuid()" json:"id"`
	AppID           string            `bun:"app_id,notnull" json:"appId"`
	AppName         string            `bun:"app_name,notnull,default:''" json:"appName"`
	SourceType      string            `bun:"source_type,notnull" json:"sourceType"`
	SourceID        string            `bun:"source_id,notnull" json:"sourceId"`
	Title           string            `bun:"title,notnull,default:''" json:"title"`
	Content         string            `bun:"content,notnull" json:"content"`
	SearchText      string            `bun:"search_text,notnull,default:''" json:"searchText"`
	Metadata        map[string]string `bun:"metadata,type:jsonb,notnull,default:'{}'" json:"metadata"`
	ContentHash     string            `bun:"content_hash,notnull" json:"contentHash"`
	SourceUpdatedAt time.Time         `bun:"source_updated_at,nullzero" json:"sourceUpdatedAt"`
	IndexedAt       time.Time         `bun:"indexed_at,notnull,default:current_timestamp" json:"indexedAt"`
}
